tests/components/bang_olufsen/test_media_player.py

```diff
@@ -1127,3 +1204,133 @@ async def test_async_browse_media(
+    expand_count: int,
+    join_count: int,
+) -> None:
+    """Test async_join_players."""
+
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+
+    source_change_callback = (
+        mock_mozart_client.get_source_change_notifications.call_args[0][0]
+    )
+
+    # Add another entity
+    mock_config_entry_2.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry_2.entry_id)
+
+    # Set the source to a beolink expandable source
+    source_change_callback(BangOlufsenSource.TIDAL)
+
+    await hass.services.async_call(
+        "media_player",
+        "join",
+        {
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_GROUP_MEMBERS: group_members,
+        },
+        blocking=True,
+    )
+
+    assert mock_mozart_client.post_beolink_expand.call_count == expand_count
+    assert mock_mozart_client.join_latest_beolink_experience.call_count == join_count
+
+
+@pytest.mark.parametrize(
+    ("source", "group_members", "expected_result", "error_type"),
+    [
+        # Invalid source
+        (
+            BangOlufsenSource.LINE_IN,
+            [TEST_MEDIA_PLAYER_ENTITY_ID_2],
+            pytest.raises(HomeAssistantError),
+            "invalid_source",
+        ),
+        # Invalid media_player entity
+        (
+            BangOlufsenSource.TIDAL,
+            [TEST_MEDIA_PLAYER_ENTITY_ID_3],
+            pytest.raises(ServiceValidationError),
+            "invalid_grouping_entity",
+        ),
+    ],
+)
+async def test_async_join_players_invalid(
+    hass: HomeAssistant,
+    mock_mozart_client: AsyncMock,
+    mock_config_entry: MockConfigEntry,
+    mock_config_entry_2: MockConfigEntry,
+    source: Source,
+    group_members: list[str],
+    expected_result: Callable,
+    error_type: str,
+) -> None:
+    """Test async_join_players with an invalid media_player entity."""
+
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+
+    source_change_callback = (
+        mock_mozart_client.get_source_change_notifications.call_args[0][0]
+    )
+
+    mock_config_entry_2.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry_2.entry_id)
+
+    source_change_callback(source)
+
+    with expected_result as exc_info:
+        await hass.services.async_call(
+            "media_player",
+            "join",
+            {
+                ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+                ATTR_GROUP_MEMBERS: group_members,
+            },
+            blocking=True,
+        )
+
+    assert exc_info.value.translation_domain == DOMAIN
+    assert exc_info.value.translation_key == error_type
+    assert exc_info.errisinstance(HomeAssistantError)
+
+    assert mock_mozart_client.post_beolink_expand.call_count == 0
+    assert mock_mozart_client.join_latest_beolink_experience.call_count == 0
+
+
+async def test_async_unjoin_player(
```

Comment:
Maybe add a negative case to assert what happens when trying to unjoin a player which is not in a group?

Reply:
The unjoin service simply calls `post_beolink_leave` regardless if the device is in a group or not, which leaves the session or does nothing, so I don't think there is anything left to test here.